project/Version.scala

```diff
@@ -0,0 +1,31 @@
+package akka
+
+import sbt._
+import sbt.Keys._
+
+/**
+ * Generate version.conf and akka/Version.scala files based on the version setting.
+ */
+object Version {
+
+  def versionSettings: Seq[Setting[_]] = inConfig(Compile)(Seq(
+    resourceGenerators <+= generateVersion(resourceManaged, _ / "version.conf",
+      """|akka.version = "%s"
+         |"""),
+    sourceGenerators <+= generateVersion(sourceManaged, _ / "akka" / "Version.scala",
+      """|package akka
+         |
+         |object Version {
+         |  val current: String = "%s"
```

Comment:
Sounds like a good reason to avoid the constant :)

-- 
Cheers,
√
On Feb 12, 2016 11:47 PM, "Peter Vlugter" <notifications@github.com> wrote:

> In project/Version.scala
> <https://github.com/akka/akka/pull/19753#discussion_r52808890>:
>
> > +import sbt.Keys._
> > +
> > +/**
> > + * Generate version.conf and akka/Version.scala files based on the version setting.
> > + */
> > +object Version {
> > +
> > +  def versionSettings: Seq[Setting[_]] = inConfig(Compile)(Seq(
> > +    resourceGenerators <+= generateVersion(resourceManaged, _ / "version.conf",
> > +      """|akka.version = "%s"
> > +         |"""),
> > +    sourceGenerators <+= generateVersion(sourceManaged, _ / "akka" / "Version.scala",
> > +      """|package akka
> > +         |
> > +         |object Version {
> > +         |  val current: String = "%s"
>
> I just tried out the constant. Certainly becomes a string literal in
> ActorSystem$.
>
> Looks like the only gotcha is incremental compiles, which don't pick up
> changes. So this:
>
> sbt> project akka-actor
> sbt> compile
> sbt> set version := "1.2.3"
> sbt> compile
> [info] Compiling 1 Scala source to ...
>
> Only the Version.scala file is recompiled. Leading to:
>
> scala> akka.Version.current
> res0: String = 1.2.3
>
> scala> ConfigFactory.load.getString("akka.version")
> res1: String = 1.2.3
>
> scala> akka.actor.ActorSystem.Version
> res2: String = 2.4-SNAPSHOT
>
> scala> startSystem()
> akka.ConfigurationException: Akka JAR version [2.4-SNAPSHOT] does not match the provided config version [1.2.3]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/19753/files#r52808890>.
>